### ISSUE TITLE
C6.4c: source-polynomial representation v3 and second untouched adequacy boundary

### ISSUE THREAD

**Issue by Alajmah** [ElephantRock/Continuity-Aware-Distributed-Inference#122]
Third bounded representation-revision slice under #102, following C6.4b PR #119 / merge `291fa8f9bd5c7d0e72e46163715cb04bd772d205`.

## Goal

Replace representation-v2 request-level piecewise interpolation with a compact source-semantic representation that directly preserves the pinned Vidur linear-regression predictor's learned polynomial component models and request-composition rules, while freezing a second genuinely untouched timing-free adequacy boundary before any new source-model reference value is observed.

C6.4c is a **contract/freeze slice only**. It must not evaluate new source-model timings and must not compute a new adequacy decision.

C6.4d will compile the frozen source models into representation-v3 profiles and evaluate the boundary frozen here. C6.4e ordinary simulator integration remains blocked until C6.4d passes.

## Why another representation revision is justified independently of C6.4b residuals

The reason to revise is the C6.4b failure, but the replacement form must not be selected from those residuals. Its form is instead dictated by the already-pinned upstream implementation at `microsoft/vidur@abae7f63aa857300f5cdc6f5e0d27860cd24721b`:

- `LinearRegressionExecutionTimePredictor` is `PolynomialFeatures` + `LinearRegression`, with upstream grid search over degree/include-bias/interaction-only/fit-intercept.
- compute components are trained against `num_tokens` and materialized on integer token keys;
- cold-prefill attention is trained against `[kv_cache_size, prefill_chunk_size_squared]`;
- cold prefill at zero processed tokens uses `kv_cache_size=0`, exact prefill chunk size, and the Batch compute-token rounding rule `(n + 7) // 8 * 8`;
- KV-cache-save uses the exact unrounded token count;
- `ExecutionTime.model_time` composes per-layer components, multiplies by the 32 Llama-2-7B layers in TP1/PP1, and converts ms to seconds;
- send/recv is a source-trained polynomial over predictor token count and is materialized only on integer keys `1..4096` under the frozen protocol.

These semantics were present before C6.4b and do not depend on the observed C6.4b error vector.

## Representation v3

Freeze representation ID:

```text
cadi.c6.cost-representation.v3.source-polynomial-composite+decode-step-affine
```

Representation-v3 profiles serialize **canonical expanded polynomial models**, not sklearn objects and not request-level source-model lookup values.

Each source polynomial must carry at least:

```text
hardware_id
source model/component identity
ordered feature names
canonical monomial powers
ordered IEEE-754 binary64 coefficients
intercept
source protocol fingerprint
training/source provenance
upstream-selected polynomial hyperparameters
```

The canonical CADI evaluator must:

- evaluate monomials in a declared deterministic order;
- combine terms with a declared deterministic summation rule;
- never clip, smooth, coerce, or force non-negative intermediate source predictions;
- fail closed on non-finite output;
- preserve the source domain rather than extrapolate beyond it.

### Cold prefill / recompute

For one cold-prefill request with `n` input tokens in the declared domain:

```text
rounded_n = (n + 7) // 8 * 8
```

The request-level model composes the pinned component polynomials exactly according to Vidur semantics:

```text
rounded_n:
  attn_pre_proj
  attn_post_proj
  attn_rope
  input_layernorm
  post_attention_layernorm
  mlp_up_proj
  mlp_down_proj
  mlp_act
  add

exact n:
  attn_kv_cache_save

attention prefill:
  features = (kv_cache_size=0, prefill_chunk_size_squared=n*n)
```

The per-layer sum is multiplied by 32 and by `1e-3` to obtain seconds. TP and PP communication remain zero in the frozen TP1/PP1 domain. CPU overhead is excluded because the accepted C6.4b protocol has `skip_cpu_overhead_modeling=True` and uses `ExecutionTime.model_time`.

Recompute uses the same cold-prefill model on the missing-prefix token count, preserving the C6.4a workload semantics.

### Decode

The single-token decode primitive remains the already-validated C6.3 affine primitive and is not refit in this slice. Multi-token decode keeps the corrected C6.4a composition:

```text
m*fixed + slope*(m*n + m*(m-1)/2)
```

No C6.4b fresh result may be used to change the decode primitive.

### Point-to-point transfer

Transfer uses the pinned upstream send/recv polynomial with the exact source mapping:

```text
predictor_token = bytes / (2 * 4096)
```

The declared calibrated transfer domain is the upstream materialized integer-key domain only:

```text
bytes = 8192 * k
1 <= k <= 4096
```

This domain is source-defined by the pinned predictor and is not chosen from C6.4b residuals. Non-multiples of 8192 and keys outside `1..4096` are outside the calibrated representation-v3 domain; no rounding or extrapolation is permitted.

### State size / memory

State-size and memory-capacity mechanics remain the existing separately sourced C6 mechanics. No new physical constants may enter C6.4c.

## Training/projection boundary

C6.4d may obtain polynomial coefficients only by executing the exact pinned Vidur training pipeline under the accepted protocol-v2 numerical substrate, then exporting the learned estimators into the canonical representation-v3 polynomial form.

Forbidden training/tuning evidence for v3:

```text
all C6.3 request-level VALIDATION points
all C6.4a/C6.4b fresh reference values
all residual/error vectors from C6.3 or C6.4b
```

The upstream raw profiling rows used by Vidur's own source training remain P-SRC2 calibration evidence and are permitted because representation v3 transcribes the source predictor rather than fitting a CADI request-level approximation to held-out results.

## Second untouched boundary

C6.4c must freeze exact second-boundary axes before any new source-model timing is evaluated.

Axis selection must be deterministic and timing-blind:

1. build the candidate identity universe from the source-declared representation-v3 domain;
2. remove every C6.3 request-level raw-reference axis for the same family;
3. remove every previously frozen/observed C6.4a boundary axis;
4. partition the remaining domain into four equal-width axis strata;
5. rank candidates inside each stratum by SHA-256 of
   `cadi.c6.4c.boundary.v1|291fa8f9bd5c7d0e72e46163715cb04bd772d205|<family>|<axis>`;
6. select the four lowest hashes per stratum;
7. serialize the final 16 axes per family in ascending numeric order.

For transfer, stratum/ranking is performed on the integer predictor-token key `k`; the serialized public axis is `8192*k` bytes.

The boundary-generation workflow may inspect **identity columns only** from the old source/reference corpus. It must not read or print source timing columns, C6.4b fresh timings, residuals, or aggregate metrics.

The frozen boundary artifact must explicitly serialize:

```text
contains_reference_timings = false
selection_algorithm_id
selection_seed
excluded-history identities/fingerprints
candidate-domain identity
16 prefill axes
16 transfer axes
```

## Fresh oracle for C6.4d

Reuse the frozen C6.3 adequacy oracle unchanged:

```text
MAPE <= 0.05
max APE <= 0.10
zero-reference points require exact zero absolute error
8-ULP decimal-threshold boundary budget
```

C6.4d must compare the compiled representation-v3 request-level outputs against a separately executed exact pinned Vidur source model at only the C6.4c frozen axes. New references remain:

```text
SIMULATED_SOURCE_MODEL_DERIVED_P_SRC2
```

They are not physical GPU measurements or EV4 evidence.

## Anti-leakage / fail-closed rules

C6.4c fails review if any of the following occurs:

- any C6.4b fresh timing or residual selects representation form, degree, coefficient, domain, or boundary axis;
- any old validation/fresh point becomes a polynomial training target;
- request-level lookup values are serialized as the representation instead of independently derived source polynomial coefficients;
- the source transfer domain is extrapolated or silently rounded;
- boundary selection reads timing columns or uses a human-selected exception after hashing;
- boundary axes are changed after any new source-model value is evaluated;
- C6.4d timing evaluation begins before the C6.4c contract/boundary merge is complete.

## Exit

Close C6.4c only after:

- representation-v3 schema and deterministic evaluator semantics are implemented;
- source-semantic cold-prefill/recompute and transfer composition contracts are regression tested;
- the carried decode contract remains unchanged;
- exact timing-free second-boundary axes are generated solely from source/history identities and committed;
- tests prove no historical C6.3/C6.4a axis enters the new boundary;
- tests prove the boundary artifact contains no reference timings;
- bounded technical/scientific review passes;
- exact-head deterministic CI and SHA-fenced squash merge complete.

No new source-model timing may be evaluated before this issue closes. No H4/H5/H7 claim is adjudicated in this slice.

**Comment by Alajmah** [comment]
C6.4c implementation is now in PR #123 from branch `c6/source-polynomial-v3-boundary`.

Timing-blind boundary selector run `34260930112` passed without evaluating any new source-model reference value. Frozen second boundary:

```text
prefill input tokens
85,202,475,897,1026,1333,1622,1623,2371,2373,2476,2636,3352,3353,3402,3899

transfer bytes
2482176,2637824,3620864,4513792,8527872,9314304,14540800,14729216,19251200,19611648,23420928,23609344,26705920,27295744,32391168,32825344
```

Committed artifact `artifacts/c6.4c/second-fresh-boundary.json` has `contains_reference_timings=false` and is byte-reproduced by the identity-only selector.

PR head `c2189da38bc611e226dceb263bf7cdad9ae4932f` currently has dedicated contract run `34261652982` green (11 tests + byte-identical boundary regeneration) and deterministic-core run `34261755643` green on Python 3.11/3.12/3.13. Automated exact-head review is in progress. C6.4d timing evaluation remains blocked until #123 merges.

**Comment by Alajmah** [comment]
Pre-merge C6.4c freeze candidate is now `9e7f74c6f5d53e97f96967793d777c976b527609` on PR #123. Exact-head dedicated run `34266127030` passes 12 contract tests, pinned-source/blob fences, and byte-identical timing-free boundary reproduction; deterministic-core run `34266132409` passes Python 3.11/3.12/3.13. The candidate additionally fails closed on non-Llama-2-7B/non-A100-H100 profile bindings and requires the exact four canonical Vidur grid-search hyperparameter keys/encodings in exported polynomial records. Final exact-head Codex review has been requested. No C6.4d boundary timing has been evaluated; the scientific freeze remains intact until merge.
